src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -198,22 +203,182 @@ std::optional<Values> Visitor::visitTypesafe(Parser::ValuesClauseContext* ctx) {
+    return blankNode.toSparql();
+  };
+  auto literal = [](const Literal& literal) {
+    // Problem: ql:contains-word causes the " to be stripped.
+    // TODO: Move stripAndLowercaseKeywordLiteral out to this point or
+    //  rewrite the Turtle Parser s.t. this code can be integrated into the
+    //  visitor. In this case the turtle parser should output the
+    //  corresponding modell class.
+    try {
+      return TurtleStringParser<TokenizerCtre>::parseTripleObject(
+          literal.toSparql());
+    } catch (const TurtleStringParser<TokenizerCtre>::ParseException&) {
+      return TripleComponent{literal.toSparql()};
+    }
+  };
+  auto graphTerm = [&iri, &blankNode, &literal](const GraphTerm& term) {
+    return term.visit(
+        ad_utility::OverloadCallOperator{iri, blankNode, literal});
+  };
+  auto varTriple = [](const Variable& var) {
+    return TripleComponent{var.name()};
+  };
+  auto varOrTerm = [&varTriple, &graphTerm](VarOrTerm varOrTerm) {
+    return varOrTerm.visit(
+        ad_utility::OverloadCallOperator{varTriple, graphTerm});
+  };
+
+  auto varPath = [](const Variable& var) {
+    return PropertyPath::fromVariable(var);
+  };
+  auto path = [](const PropertyPath& path) { return path; };
+  auto varOrPath = [&varPath,
+                    &path](ad_utility::sparql_types::VarOrPath varOrPath) {
+    return std::visit(ad_utility::OverloadCallOperator{varPath, path},
+                      std::move(varOrPath));
+  };
+
+  auto registerIfVariable = [this](auto& variant) {
+    if (holds_alternative<Variable>(variant)) {
+      visibleVariables_.back().emplace_back(std::get<Variable>(variant));
+    }
+  };
+
+  auto convertTriple = [&varOrTerm, &varOrPath, &registerIfVariable](
+                           TripleWithPropertyPath&& triple) -> SparqlTriple {
+    registerIfVariable(triple.subject_);
+    registerIfVariable(triple.predicate_);
+    registerIfVariable(triple.object_);
+
+    return {varOrTerm(std::move(triple.subject_)),
+            varOrPath(std::move(triple.predicate_)),
+            varOrTerm(std::move(triple.object_))};
+  };
+
+  vector<SparqlTriple> triples = ad_utility::transform(
+      ad_utility::flatten(
+          visitVector<vector<TripleWithPropertyPath>>(ctx->triplesBlock())),
+      convertTriple);
+  auto [ops, filters] =
+      splitUp(visitVector<std::variant<GraphPatternOperation, SparqlFilter>>(
+          ctx->graphPatternNotTriples()));
+  ops.emplace_back(
+      GraphPatternOperation::BasicGraphPattern{std::move(triples)});
+  return {ops, filters};
+}
+
 // ____________________________________________________________________________________
 vector<TripleWithPropertyPath> Visitor::visitTypesafe(
     Parser::TriplesBlockContext* ctx) {
   auto triples = visitTypesafe(ctx->triplesSameSubjectPath());
   if (ctx->triplesBlock()) {
-    appendVector(triples, visitTypesafe(ctx->triplesBlock()));
+    ad_utility::appendVector(triples, visitTypesafe(ctx->triplesBlock()));
```

Comment:
Currently, this is a little bit ineffective due to the way the rule is written.
Mabybe (by using a helper function etc.) we could first get a `vector<vector<Triple>>` where each of the recursion just appends, and then at the end we concatenate/flatten them all.
(Typically this shouldn't be an issue, but out of principle we should write the "correct" code considering performance).